### PR TITLE
Introduction of Internal Shift Variant for New Gene Induction

### PR DESCRIPTION
**Pull request by rjuenemann**
A new internal shift variant framework is introduced. This allows users to define variants that change `sim_data` at the beginning of specified generation numbers. For more details on how to adapt this for a particular use case, refer to `docs/misc/internal_shift.md`.

The internal shift framework defines a dictionary in `sim_data` specifying which modification functions to apply at each generation. Codebase changes were made to the daughter cell initiation process to keep track of the generation number and apply the modifications before beginning that cell’s simulation. For these modified values to persist and be used in the simulation, PR #1395 was necessary to move the initialization of processes after setting initial conditions. We chose to approach it in this way for ease of implementation (including consistency with existing variant pipelines), generalizability to multiple shifts and use cases outside of new genes, and accessibility of parameter choices for analysis scripts.

Introducing the internal shift variant framework allows us to resolve Issue #1376 and build upon PR #1369, making it possible to induce and knock out the expression of new genes added to the chromosome at the start of chosen generation numbers. This enables the analysis of the transitory cellular response when new gene expression is initially induced, as well as stabilizes some failures we were seeing as artifacts of initialization in the first generation for high new gene expression factors. This PR contains the model infrastructure necessary to enable these simulations, and a separate future PR will focus on the analysis scripts created to analyze them.

Here is an example of new gene mRNA and protein counts for a 4-generation `new_gene_expression_and_translation_efficiency_internal_shift` simulation with a single _gfp_ added to the chromosome. New gene expression was induced at the beginning of generation 1 and then knocked out in generation 2.

[new_gene_counts.pdf](https://github.com/CovertLab/wcEcoli/files/13590298/new_gene_counts.pdf)

### CODE REVIEW

File: models/ecoli/sim/variants/new_gene_expression_and_translation_efficiency_internal_shift.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+	shortName = "control",
+	desc = "Control simulation"
+	)
+
+# NOTE: The version of these global variables on the master branch will
+# be used for Jenkins testing
+
+NEW_GENE_INDUCTION_GEN = 1 # Generation to induce new gene expression
+NEW_GENE_KNOCKOUT_GEN = 2 # Generation to knock out new gene expression
+FINAL_SHIFT_GEN = 128 # Generation to stop knocking out new gene expression
```
Comment (ggsun) [inline models/ecoli/sim/variants/new_gene_expression_and_translation_efficiency_internal_shift.py:53]: Is "stop knocking out" equivalent to reinducing?

File: models/ecoli/sim/variants/new_gene_expression_and_translation_efficiency_internal_shift.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+	return expression_factor, trl_eff_value
+
+
+def induce_new_genes(sim_data, index):
+	"""
+	'Induces' new genes by setting their expression levels and translation
+	effiencies to the values specified using the variant index.
+	"""
+	# Map variant index to expression factor and tranlsation efficiency value
```
Comment (ggsun) [inline models/ecoli/sim/variants/new_gene_expression_and_translation_efficiency_internal_shift.py:141]: typo in tranlsation

File: models/ecoli/sim/variants/template_internal_shift.py
Comment (ggsun) [inline models/ecoli/sim/variants/template_internal_shift.py:1]: Good call on adding this template file!

**Review by ggsun — APPROVED** [review]
Looks good to me! Just some minor comments on some things that were a bit unclear to me on the first read.

**Review by ggsun — APPROVED** [review]

File: models/ecoli/sim/initial_conditions.py [hunk 3]
```diff
@@ -115,3 +115,15 @@ def setDaughterInitialConditions(sim, sim_data):
 	sim.internal_states["UniqueMolecules"].loadSnapshot(inherited_state['unique_molecules'])
 
 	sim._initialTime = inherited_state['initial_time']
+	sim._generationNumber = inherited_state['generation_number']
```
[the remark this answers is not in the document]
Comment (rjuenemann) [reply models/ecoli/sim/initial_conditions.py]: I agree - thanks for catching this!

**Review by ggsun — APPROVED** [review]

**Pull request merged**